### PR TITLE
Kuryr CNI support

### PR DESCRIPTION
**Pull request by yuvalif**
This PR allows the VM to get multiple networks/interfaces via Kuryr CNI.

_**Functional tests are missing.**_

**Release Notes:**
To add multiple networks/interfaces to a Virtual Machine Instance, add the following network definitions:
```yaml
networks:
  - kuryr:
      networkName: net-a-conf
    name: net-a
```
Where, ```networkName``` is the name of the ```NetworkAttachmentDefinition``` CRD to be used. Kuryr does not delegate the network definition to another CNI, and will provide the networking based on an existing openstack networking solution which has to be deployed beforehand. The parameters of the network will be sent in the annotation part of the ```NetworkAttachmentDefinition``` CRD. For example:
```yaml
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: net-a-conf
  annotations:
    openstack.org/kuryr-config: '{
      "subnetId": "id_of_neutron_subnet_created_previously"
    }'
```
> Notes: 
> * Only one: Multus, Genie or Kuryr could be used across all networks
> * Pod network could be combined with Kuryr, where it will get ```eth0``` in the pod without any ```NetworkAttachment``` CRD. The network definition will be taken from the default of Kuryr as configuration under ```/etc/cni/net.d/``` 

```release-note
see release notes above
```

### CODE REVIEW

**Comment by yuvalif** [comment]
no functional tests
/hold

**Comment by yuvalif** [comment]
no functional tests
/hold

**Comment by yuvalif** [comment]
[Kuryr: add support](https://trello.com/c/dgUfL4VN/242-kuryr-add-support)

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook.go [hunk 1]
```diff
@@ -821,6 +822,12 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 				networkNameExistsOrNotNeeded = network.Genie.NetworkName != ""
 			}
 
+			if network.NetworkSource.Kuryr != nil {
+				cniTypesCount++
+				kuryrExists = true
+				networkNameExistsOrNotNeeded = network.Kuryr.NetworkName != ""
```
Comment (phoracek) [inline pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: not a part of this PR, but this variable does not make sense with the error message "CNI delegating plugin must have a networkName"

File: pkg/virt-launcher/virtwrap/network/network.go [hunk 2]
```diff
@@ -76,6 +79,9 @@ func SetupNetworkInterfaces(vmi *v1.VirtualMachineInstance, domain *api.Domain)
 		} else if networks[iface.Name].Genie != nil {
 			// genie pod interfaces named ethX
 			podInterfaceName = fmt.Sprintf("eth%d", cniNetworks[iface.Name])
+		} else if networks[iface.Name].Kuryr != nil {
+			// genie pod interfaces named ethX
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/network.go]: s/genie/kuryr/

File: pkg/virt-api/webhooks/validating-webhook/validating-webhook.go (hunk 1 above)
Comment (yuvalif) [reply pkg/virt-api/webhooks/validating-webhook/validating-webhook.go]: why? this variable is "true" in one of two cases: network name exists, or not needed.
if the network name is needed, but not set, we gave the above error

**Comment by fabiand** [comment]
Please either update or close this PR

**Comment by yuvalif** [comment]
@slintes, I don't know what is the priority of the feature. what is blocking it is lack of integration testing.
In order to achieve that we should add a k8s-kuryr image to our standard CI (probably based on devstack, see: https://docs.openstack.org/kuryr-kubernetes/latest/installation/devstack/index.html). 
If someone could prepare that, I would be happy to finish the feature, doing rebase, and writing the actual tests.

**Comment by slintes** [comment]
@yuvalif thanks!
@fabiand can you say something about priority?

**Comment by fabiand** [comment]
For me it has a low priority.

**Comment by fabiand** [comment]
I actually wonder - with the npwgv1 support, kuryr should be supported if it's upporting the network standard, not? @booxter

**Comment by yuvalif** [comment]
@fabiand kuryr is not a network definition delegation framework. It is the network provider itself (for multiple networks). This is why the information there is encoded in differently in the CRD.

**Comment by fabiand** [comment]
Right. I just wondered if they were catching up with it and provided
support for that standard as well.

**Pull request closed without merging**